ImputationPipeline/ScoringTasks.wdl

```diff
@@ -199,6 +199,123 @@ task AddInteractionTermsToScore {
   }
 }
 
+task CheckWeightsCoverSitesUsedInTraining {
+  input {
+    File sites_used_in_training
+    WeightSet weight_set
+  }
+
+  command <<<
+    python3 << "EOF"
+    import csv
+    import sys
+
+    with open("~{sites_used_in_training}") as f_sites_used_in_training:
+      sites_used_in_training = {s.strip() for s in f_sites_used_in_training}
+
+    sites_in_weight_set = set()
+    with open("~{weight_set.linear_weights}") as f_linear_weights:
+      linear_weights_reader = csv.reader(f_linear_weights, delimiter='\t')
+      next(linear_weights_reader)
+      for line in linear_weights_reader:
+        sites_in_weight_set.add(line[0])
+
+    if ~{if (defined(weight_set.interaction_weights)) then "True" else "False"}:
+      with open("~{if (defined(weight_set.interaction_weights)) then select_first([weight_set.interaction_weights]) else ''}") as f_interaction_weights:
+        for line in csv.DictReader(f_interaction_weights, delimiter='\t'):
+          sites_in_weight_set.add(line['id_1'])
+          sites_in_weight_set.add(line['id_2'])
+
+    sites_missing_from_weight_set = sites_used_in_training - sites_in_weight_set
+
+    if len(sites_missing_from_weight_set) > 0:
+      sys.exit(f"Error: {len(sites_missing_from_weight_set)} sites used in model training are missing from weights files.")
+    EOF
+
+
+  >>>
+
+  runtime {
+    docker : "python:3.9.10"
+  }
+}
+
+task CompareScoredSitesToSitesUsedInTraining {
+  input {
+    File sites_used_in_training
+    File sites_used_in_scoring
+    WeightSet weight_set
+  }
+
+  command <<<
+    python3 << "EOF"
+    import csv
+
+    with open("~{sites_used_in_training}") as f_sites_used_in_training:
+      sites_used_in_training = {s.strip() for s in f_sites_used_in_training}
+
+    with open("~{sites_used_in_scoring}") as f_sites_used_in_scoring:
+      sites_used_in_scoring = {s.strip() for s in f_sites_used_in_scoring}
+
+    missing_sites = sites_used_in_training - sites_used_in_scoring
+
+    with open("missing_sites.txt", "w") as f_missing_sites:
+      for site in missing_sites:
+        f_missing_sites.write(site)
+
+    with open("n_missing_sites.txt", "w") as f_n_missing_sites:
+      f_n_missing_sites.write(f"{len(missing_sites)}")
+
+    max_error_up = 0
+    max_error_down = 0
+
+    with open("~{weight_set.linear_weights}") as f_weights:
+      weights_reader = csv.reader(f_weights, delimiter = "\t")
+      next(weights_reader)
+      for line in weights_reader:
+        id = line[0]
+        if id in missing_sites:
+          missing_site_weight = float(line[2])
+          if missing_site_weight > 0:
+            max_error_up += 2 * missing_site_weight
+          else:
+            max_error_down += 2 * missing_site_weight
+
+
+    if ~{if defined(weight_set.interaction_weights) then "True" else "False"}:
+      with open("~{if (defined(weight_set.interaction_weights)) then select_first([weight_set.interaction_weights]) else ''}") as f_interaction_weights:
```

Comment:
Same here

Reply:
You would think, but the wdl spec calls for a runtime workflow failure is a select_first is evaluated  with no defined inputs.  However, this entire ugly if/then/select_first clause can just be replaced by `select_first([weight_set.interaction_weights, ''])`, which is definitely what I should have done in the first place, lol

Reply:
what am I talking about, I don't even need the selec_first, it will just evaluate to an empty string if it's not defined anyway.  oy vey, long week